packages/backend/src/core/VerifierCollector.ts

```diff
@@ -0,0 +1,111 @@
+import { utils } from 'ethers'
+import { AbiCoder } from 'ethers/lib/utils'
+
+import { EthereumAddress } from '../model'
+import {
+  ImplementationAddedEventRecord,
+  UpgradedEventRecord,
+  VerifierEventRecord,
+  VerifierEventRepository,
+} from '../peripherals/database/VerifierEventRepository'
+import { EthereumClient } from '../peripherals/ethereum/EthereumClient'
+import { BlockRange } from '../peripherals/ethereum/types'
+import { partition } from '../tools/partition'
+
+const PROXY_ADDRESS = '0xC8c212f11f6ACca77A7afeB7282dEBa5530eb46C'
+
+const PROXY_ABI = new utils.Interface([
+  'event ImplementationAdded(address indexed implementation, bytes initializer, bool finalize)',
+  'event Upgraded(address indexed implementation)',
+])
+
+const Upgraded = PROXY_ABI.getEventTopic('Upgraded')
+const ImplementationAdded = PROXY_ABI.getEventTopic('ImplementationAdded')
+
+export class VerifierCollector {
+  constructor(
+    private readonly ethereumClient: Pick<EthereumClient, 'getLogs'>,
+    private readonly verifierEventRepository: VerifierEventRepository
+  ) {}
+
+  /**
+   * Saves new `Upgraded` and `ImplementationAdded` events to the database,
+   * and returns verifier addresses derived from both the old and new events.
+   */
+  async collect(blockRange: BlockRange): Promise<EthereumAddress[]> {
+    const oldEvents = this.verifierEventRepository.getAll()
```

Comment:
I though that we should only query events up to this point, so `.getAllBefore(blockRange)`, but because we are guaranteed to have no newer events then getAll may be fine.

Reply:
Should VerifierCollector have a `revert` or `cleanup` method then? We'd call it in DataSyncService `.revert` and remove all events after block number.